### PR TITLE
Update to Sequelize 2.0.2

### PR DESCRIPTION
**Pull request by fduch2k**

### CODE REVIEW

**Comment by bebraw** [comment]
:+1:

Getting this merged would be great.

File: lib/passport-local-sequelize.js [hunk 1]
```diff
@@ -62,75 +67,83 @@ var defaultUserSchema = {
...
-                    return cb(err);
+                    reject(err);
                 }
 
-                self.set(options.hashField, new Buffer(hashRaw, 'binary').toString('hex'));
-                self.set(options.saltField, salt);
+                var salt = buf.toString('hex');
 
-                cb(null, self);
+                crypto.pbkdf2(password, salt, options.iterations, options.keylen, options.digest, function (err, hashRaw) {
```
Comment (bebraw) [inline lib/passport-local-sequelize.js]: This fails at Node 0.10 as the function signature has changed.

* **0.10** - `crypto.pbkdf2(password, salt, iterations, keylen, callback)`
* **0.12** - `crypto.pbkdf2(password, salt, iterations, keylen[, digest], callback)`

File: test/all_spec.js [hunk 2]
```diff
@@ -73,12 +68,13 @@ describe('Passport Local Sequelize', function () {
...
 
                     authenticatedUser.should.not.equal(false);
-
                     authenticatedUser.get('username').should.equal('someuser');
-
                     done();
                 });
             });
+        }).catch(function (error) {
+            console.log(1);
```
Comment (bebraw) [inline test/all_spec.js]: Might want to drop this `console.log`.

File: lib/passport-local-sequelize.js [hunk 3]
```diff
@@ -62,75 +67,83 @@ var defaultUserSchema = {
...
-            cb(null, self);
-          
         });
     };
 
     UserSchema.DAO.prototype.authenticate = function (password, cb) {
         var self = this;
         // TODO: Fix callback and behavior to match passport
-        crypto.pbkdf2(password, this.get(options.saltField), options.iterations, options.keylen, function (err, hashRaw) {
+        crypto.pbkdf2(password, this.get(options.saltField), options.iterations, options.keylen, options.digest, function (err, hashRaw) {
```
Comment (bebraw) [inline lib/passport-local-sequelize.js]: Same signature here as above.

File: lib/passport-local-sequelize.js [hunk 4]
```diff
@@ -165,65 +178,49 @@ var attachToUser = function (UserSchema, options) {
...
-            user = self.build(fields);
-        } else if (_.isObject(user)) {
-            // Create an instance if user is passed as fields
-            user = self.build(user);
-        }
 
-        if (!user.get(options.usernameField)) {
-            return cb(new Error(util.format(options.missingUsernameError, options.usernameField)));
-        }
+    UserSchema.register = function (user, password) {
```
Comment (bebraw) [inline lib/passport-local-sequelize.js]: This should use a signature like `function (user, password, cb) {` instead to be consistent with other `passport-local` libraries like [passport-local-mongoose](https://www.npmjs.com/package/passport-local-mongoose).

Incidentally I'm trying to replace that library with the Sequelize version. :)

If you want to retain the promisified version, perhaps something can be done about that? Ie. if callback is provided, rely on that, else return a promise. Then you get something that works with your way too.

Comment (fduch2k) [inline lib/passport-local-sequelize.js]: I'll rollback this changes. Currently I'm using promisify(register) to get promise version

Comment (bebraw) [inline lib/passport-local-sequelize.js]: Alright, I think overall it's really close. :+1:

I'm not sure what's the cleanest way to deal with those `crypto.pbkdf2` bits. Maybe you need to wrap that and detect for Node version?

Comment (fduch2k) [inline lib/passport-local-sequelize.js]: I'll think about crypto.pbkdf2

**Comment by madhurjain** [comment]
@bebraw @fduch2k Thanks for your contributions, I will test this locally before I merge and push to npm

**Comment by ieugen** [comment]
+1

File: lib/passport-local-sequelize.js (hunk 4 above)
Comment (fduch2k) [inline lib/passport-local-sequelize.js]: @bebraw I changed my mind about callback signatures. Here is my thoughts: this lib extends sequelize model, so it's more logically follow sequelize conventions but not passport-local-mongoose. I restored callback signatures for this PR and start new branch to continue port to pomise style

File: lib/passport-local-sequelize.js [hunk 5]
```diff
@@ -86,8 +99,7 @@ var attachToUser = function (UserSchema, options) {
             }
 
             var salt = buf.toString('hex');
-
-            crypto.pbkdf2(password, salt, options.iterations, options.keylen, function (err, hashRaw) {
+            shimPbkdf2(password, salt, options.iterations, options.keylen, options.digest, function (err, hashRaw) {
```
Comment (madhurjain) [inline lib/passport-local-sequelize.js]: Given that the digest paramter is optional ([docs](https://nodejs.org/api/crypto.html#crypto_crypto_pbkdf2_password_salt_iterations_keylen_digest_callback)) defining a `shimPbkdf2` method and checking the node version is overkill

Comment (bebraw) [inline lib/passport-local-sequelize.js]: @madhurjain I recall the problem here was that Node API for crypto changed between 0.10 and 0.12. Of course if it's possible to simplify this, great!

Comment (fduch2k) [inline lib/passport-local-sequelize.js]: nodejs 0.10.x does not supported digest param at all - ([docs](https://nodejs.org/docs/v0.10.0/api/crypto.html#crypto_crypto_pbkdf2_password_salt_iterations_keylen_callback)). That's why I wrap crypto.pbkdf2 by shimPbkdf2

File: .jshintrc [hunk 6]
```diff
@@ -2,6 +2,7 @@
   "node": true,
   "browser": true,
   "esnext": true,
+  "predef": [ "-Promise" ],
```
Comment (madhurjain) [inline .jshintrc]: from previous commit and no more required ?

File: package.json [hunk 7]
```diff
@@ -23,8 +23,9 @@
   ],
   "dependencies": {
     "passport-local": "~1.0.0",
-    "sequelize": "~1.7.2",
-    "lodash": "~2.4.1"
+    "sequelize": "~2.0.3",
+    "bluebird": "~2.4.2",
```
Comment (madhurjain) [inline package.json]: from previous commit and no more required ?

File: lib/passport-local-sequelize.js [hunk 8]
```diff
@@ -165,65 +181,65 @@ var attachToUser = function (UserSchema, options) {
...
-                        if (err) { return cb(err); }
-                        cb(null, user);
+
+                    user.setActivationKey(function (err, user) {
+
+                        if (err) {
+                            return cb(err);
+                        }
+
+                        user.save().complete(function (err) {
```
Comment (madhurjain) [inline lib/passport-local-sequelize.js]: All methods return a `Promise` object and `complete` is not supported

**Comment by madhurjain** [comment]
@fduch2k I greatly appreciate your efforts on this. But, given the time it took me to test this locally, the API seems to have changed so I cannot merge this pull request. I spent some time fixing things myself to get this library working with the latest (`2.1.3`) version of Sequelize. I've pushed it to `npm`

**Pull request closed without merging**